File: 2022/11_monkey_in_the_middle/part1.py
monkeys = {
    "monkey0": {
        "count": 0,
        "items": [52, 78, 79, 63, 51, 94],
        "op": lambda old: old * 13,
        "divisor": 5,
        "true": "monkey1",
        "false": "monkey6",
    },
    "monkey1": {
        "count": 0,
        "items": [77, 94, 70, 83, 53],
        "op": lambda old: old + 3,
        "divisor": 7,
        "true": "monkey5",
        "false": "monkey3",
    },
    "monkey2": {
        "count": 0,
        "items": [98, 50, 76],
        "op": lambda old: old * old,
        "divisor": 13,
        "true": "monkey0",
        "false": "monkey6",
    },
    "monkey3": {
        "count": 0,
        "items": [92, 91, 61, 75, 99, 63, 84, 69],
        "op": lambda old: old + 5,
        "divisor": 11,
        "true": "monkey5",
        "false": "monkey7",
    },
    "monkey4": {
        "count": 0,
        "items": [51, 53, 83, 52],
        "op": lambda old: old + 7,
        "divisor": 3,
        "true": "monkey2",
        "false": "monkey0",
    },
    "monkey5": {
        "count": 0,
        "items": [76, 76],
        "op": lambda old: old + 4,
        "divisor": 2,
        "true": "monkey4",
        "false": "monkey7",
    },
    "monkey6": {
        "count": 0,
        "items": [75, 59, 93, 69, 76, 96, 65],
        "op": lambda old: old * 19,
        "divisor": 17,
        "true": "monkey1",
        "false": "monkey3",
    },
    "monkey7": {
        "count": 0,
        "items": [89],
        "op": lambda old: old + 2,
        "divisor": 19,
        "true": "monkey2",
        "false": "monkey4",
    },
}

for i in range(20):
    for j in range(8):
        monkey = monkeys[f"monkey{j}"]
        for idx, item in enumerate(monkey["items"]):
            monkey["count"] += 1
            new = monkey["op"](item)
            new = int(new / 3)
            if new % monkey["divisor"] == 0:
                monkeys[monkey["true"]]["items"].append(new)
            else:
                monkeys[monkey["false"]]["items"].append(new)
        monkey["items"].clear()

counts = []
for k, v in monkeys.items():
    counts.append(v["count"])
counts.sort(reverse=True)
print(counts[0] * counts[1])

File: 2022/11_monkey_in_the_middle/part1.go
package main

import "fmt"

type monkey struct {
	count   int
	items   []int
	op      func(int) int
	divisor int
	monk_t  *monkey
	monk_f  *monkey
}

func main() {
	monkey0 := monkey{
		items:   []int{79, 98},
		op:      func(old int) int { return old * 19 },
		divisor: 23,
	}
	monkey1 := monkey{
		items:   []int{54, 65, 75, 74},
		op:      func(old int) int { return old + 6 },
		divisor: 19,
	}
	monkey2 := monkey{
		items:   []int{79, 60, 97},
		op:      func(old int) int { return old * old },
		divisor: 13,
	}
	monkey3 := monkey{
		items:   []int{74},
		op:      func(old int) int { return old + 3 },
		divisor: 17,
	}
	monkey0.monk_t = &monkey2
	monkey0.monk_f = &monkey3
	monkey1.monk_t = &monkey2
	monkey1.monk_f = &monkey0
	monkey2.monk_t = &monkey1
	monkey2.monk_f = &monkey3
	monkey3.monk_t = &monkey0
	monkey3.monk_f = &monkey1

	monkeys := []*monkey{&monkey0, &monkey1, &monkey2, &monkey3}

	for i := 0; i < 20; i++ {
		for _, mon := range monkeys {
			for _, item := range mon.items {
				mon.count++
				new := mon.op(item) / 3
				if new%mon.divisor == 0 {
					mon.monk_t.items = append(mon.monk_t.items, new)
				} else {
					mon.monk_f.items = append(mon.monk_f.items, new)
				}
			}
			mon.items = []int{}
		}
	}

	fmt.Println(monkey0.count)
	fmt.Println(monkey1.count)
	fmt.Println(monkey2.count)
	fmt.Println(monkey3.count)

}
